### PR TITLE
Grunt build

### PR DESCRIPTION
**Pull request by ricksbrown**
Closes issue #4.

To get the full benefit of this PR you need to go [here](http://docs.travis-ci.com/user/getting-started/) and perform:

[Step1](http://docs.travis-ci.com/user/getting-started/#Step-one%3A-Sign-in)
[Step 2](http://docs.travis-ci.com/user/getting-started/#Step-two%3A-Activate-GitHub-Webhook)

It is probably not obvious what this does and you're thinking WTF, this is not something I need, now I have a bunch of complexity I don't want.

It's going to give you a lot - for example I noticed you were minifying your JS on the live site, that should be performed automatically and that's the kind of thing Grunt is for.

For now I have wired up [JSHint](http://jshint.com/docs/) and nothing more. Run it and you'll see the magic.

### CODE REVIEW

**Comment by ricksbrown** [comment]
@peakcontent PTAL (Please Take A Look)

**Comment by peakcontent** [comment]
Hmm, this requires Node.js, which I haven't started using yet...?

**Comment by ricksbrown** [comment]
How were you planning to automate tasks?

**Comment by peakcontent** [comment]
Hi - do you mean tasks such as updating the table of contents to include new content (that is, from the 'users' side)? Or development tasks?

**Comment by ricksbrown** [comment]
Node is only needed for development, this does not add runtime dependency.

Already you have manual steps, looks to me like you did something like this:
1. Copy the content of script.js and paste it into some online minification tool.
2. Remove the `script` element which imports script.js and replace it with a script block that contains the output of the minification tool.

That's not a repeatable build with consistent outcomes. It's also not convenient.
Then add static analysis and unit testing.
Then add continuous integration.

If you want to see what it currently does then clone my fork and run grunt on it as per the readme, you'll see JSHint complain about all your superfluous semicolons.

`git clone https://github.com/RickSBrown/nothingCMS.git`

**Comment by peakcontent** [comment]
Yes, I see some of those warnings (below) - is there a different interface I can use to see the grunt output and go through it systematically?

At the moment the grunt operation is just applying JShint right? I could also do that in my new IDE (Brackets, based on your suggestion of choosing something cross-platform...)

[cid:fc5554be-1ed0-428c-81ec-83177d7693ff]

**Comment by ricksbrown** [comment]
You can run it from Netbeans - actually I realised there is an HTML5 distro
of netbeans https://netbeans.org/downloads/ which would be pretty useful
for you. The output would be much the same though.

Of course if you follow all my instructions then it will run automatically
on TravisCI every commit to Github. Then you can view it in the cloud, like
this: https://travis-ci.org/RickSBrown/nothingCMS

This means every commit is checked automagically - for example take a look
here: https://github.com/RickSBrown/nothingCMS/branches

See the red cross, hover over it. It should be a green tick and will be
once you fix all the grunt issues.

**Comment by peakcontent** [comment]
Okay, I think I'm getting somewhere...?

https://travis-ci.org/peakcontent/nothingCMS

**Comment by ricksbrown** [comment]
Yeah but I would avoid using comments in code to configure JSHint, it's not really a scalable way to do it.

Instead edit the jshint options in `Gruntfile.js`. For example it might bitch about console because I forgot to add [this option](http://jshint.com/docs/options/#devel):

`devel: true`

Edit: also I would screw old IE and delete any ActiveX code - it should not be needed nowadays.
If you want to keep it you'll have to add it to the `globals` [option](http://jshint.com/docs/options/#globals).

**Comment by peakcontent** [comment]
Okay, I've implemented those changes (JSHint comments removed, Gruntfile.js updated, support for old IE gone!).

What should be done about the continuing failures in the Travis CI output?

By the way, do you use a paid version of GitHub?

**Comment by ricksbrown** [comment]
Looks to me that you only have one failure left. Long story short you should stop using DOM Level 0 event model and move to DOM Level 2.

In plain English that means don't do this:

`element.onclick = someFunction;`

Do this instead:

`element.addEventListener("click", someFunction, false);`

Then your `someFunction` event listener will be called with a nice Event object and you can use its `currentTarget` or `target` property instead of `this`.

---

FYI I consider it bad practice to put tagnames in JS variables. They so often end up being a lie and confusing you. For example `toc_ul` might be a `UL` today but at some point it may change to something else, then you have to go rename all your JS variables or leave them confusing. Name them by what their role is, not their tagname.

---

You have to pay for private github repository, all mine are open source, public and free.
